### ISSUE TITLE
DisposeBag causes subscriptions to hang around longer than necessary / leak

### ISSUE THREAD

**Issue by sirlantis** [ReactiveX/RxSwift#573]
tl;dr Should DisposeBag use some kind of WeakMap mechanism to keep references?

> Using dispose bags or takeUntil operator is a robust way of making sure resources are cleaned up. We recommend using them in production even if the sequences will terminate in finite time.

After reading this in the [Documentation](https://github.com/ReactiveX/RxSwift/blob/master/Documentation/GettingStarted.md) I assumed the semantics of the DisposeBag to be "restricts lifetime of a subscription to the lifetime of another object".

Yesterday however I was confused to find that the following code would leak `HttpClient` instances when invoking `request` multiple times:

```
func request() {
  let httpClient = HttpClient()
  httpClient.performRequest().single().subscribeNext { [weak self] response in
    self?.httpClient(httpClient, didReceiveResponse:response)
  }.addDisposableTo(bag)
}
```

This happens because the bag holds a strong reference to the subscription, so even if the original `Observable` finishes, it will still keep this reference.

Workarounds for that leak:

- Bad: Use `[unowned httpClient]` and introduce a race condition
- Optimistic: Drop unnecessary `addDisposableTo` and hope that `httpClient.performRequest()` terminates
- Correct: Use `takeUntil(rx_deallocated)` instead of `addDisposableTo`

So, my question: Should DisposeBag use some kind of WeakMap mechanism to keep references?

If the answer is *Yes* I might find the time to research/implement this.

**Comment by kzaher** [comment]
Hi @sirlantis ,

I think that problem that you are experiencing is caused by a non idiomatic use of this library :)

If you are doing a lot of manual subscribe calls, like in your example, then you can run into issues like that.

The way to write that method in a more idiomatic way would be:

```swift
func request() -> Observable<(HttpClient, Response)> {
  let httpClient = HttpClient()
  return httpClient.performRequest().single()
}
```

If you use just operators, they will chain those disposables for you, so you won't need a dispose bag :)

**Comment by sirlantis** [comment]
Hey @kzaher, thanks for your reply.

I'm not sure what you mean by 'non-idiomatic use' and 'manual subscribe calls'. I blame me for not providing a better and less confusing example - I should not have had that delegate-ish looking API in there.

Here's a different example (closer to what you proposed) which is still leaking.

```swift
// class SomeController: UIController {

let bag = DisposeBag()

func requestWithTimeMeasurement() -> Observable<(AnyObject, NSTimeInterval)> {
  let requestStartedAt = NSDate() // leaks each time `buttonPressed` is called

  return api.request()
    .map { ($0, NSDate() - requestStartedAt) }
}

@IBAction func buttonPressed() {
  requestWithTimeMeasurement()
    .map { "Duration: \($1)" } // don't care about the response
    .subscribe(someLabel.rx_text)
    .addDisposableTo(bag)
}

```

**Comment by kzaher** [comment]
Hi @sirlantis ,

more idiomatic way to use the API would be to write something like

```
button.rx_tap
    .flatMapLatest { _ in
         requestWithTimeMeasurement()
              .map { "Duration: \($1)" } // don't care about the response
              .catchErrorJustReturn(something)
              .observeOn(MainScheduler.instance)
    }
    .subscribe(someLabel.rx_text)
    .addDisposableTo(bag)
```

**Comment by sergdort** [comment]
Hi guys !
@sirlantis  actually leak in your first example
```swift
let httpClient = HttpClient()
httpClient.performRequest().single().subscribeNext { [weak self] response in
       self?.httpClient(httpClient, didReceiveResponse:response)
  }
.addDisposableTo(bag)
```

I think here you got leak not because of `DisposeBag`, but because of retain cycle
since `httpClient` keeps strong reference to the `subscribeNext ` block (not directly) and you trying to access the `httpClient ` within block so block retains `httpClient `

**Comment by sergdort** [comment]
Hmm ... 
Looks like I'm wrong :)
`self ` don't have strong reference  `httpClient `

**Comment by sirlantis** [comment]
@kzaher Indeed that fixes the "leaks each time `buttonPressed` is called" issue. However one instance of `NSDate` will always hang around. Another shot on getting to the gist:

```swift
// assume that this implementation is hidden to me
func request() -> Observable<SomeBigObject> {
  return Observable.just(SomeBigObject())
}

// class SomeController: UIViewController {
// override func viewDidLoad() {

request()
  .map { $0.description }
  .subscribe(someLabel.rx_text)
  .addDisposableTo(bag)
```

My issue is that an instance of `SomeBigObject` will hang around now until the `UIViewController` deallocates.

While you could argue 'it's just one instance' if that happens to be a `UIImage` that can be quite a bit of memory (that's my use-case at the moment).

Hitting the leak multiple times can happen easily (therefore amplifying the effect of course). Imagine the above example code being called in `viewWillAppear` instead of `viewDidLoad`. Then more instances will leak every time the VC appears.

You could work around this by being super cautious (or: 'more idiomatic' / 'more pure') by adding a `rx_viewDidAppear` observable yourself, setting up the observation in `viewDidLoad` and then use any of the `_Latest` methods (as @kzaher proposed before). But even then: one instance will always survive.

I'm arguing that changing the DisposeBag to have weak references to its disposables would prevent these leaks from happening altogether.

**Comment by kzaher** [comment]
Hi @sirlantis ,

I recently had some more time to think about this.

First of all, thnx for reporting this, I think that we know how to improve this behavior, and we'll try to improve it for 3.0 version to handle those scenarios more gracefully. 

I don't think this can be solved efficiently by having `DisposeBag` store weak references. I understand why that seems to work, but that solution does have some problems.

Gist of this is:
* can you maybe use `takeUntil(object.rx_deallocated)` until 3.0 version? I think that should solve the problems you were experiencing in case you are sending big objects.
* we would want to improve this systematically for 3.0 since we'll be doing some tweaks for it anyway.

**Comment by tcurdt** [comment]
@kzaher hm ... but `takeUntil` is not always available:

        self.foo?.index
        .subscribeNext({ (newIndex) in
        })
        .takeUntil(self.rx_deallocated)

At least not in the above scenario - while this does work:

        self.foo?.index
        .subscribeNext({ (newIndex) in
        })
       .addDisposableTo(disposeBag)

**Comment by kzaher** [comment]
Hi @sirlantis ,

We've improved this behavior in 3.0.0.

The improvement is that once sequence terminates, it will detach all internal resources from returned subscription `Disposable`s.

Dispose bags still work like they have in 2.0.0. If one wants to create interfaces similar to this one
```swift
func request() {
  let httpClient = HttpClient()
  httpClient.performRequest().single().subscribeNext { [weak self] response in
    self?.httpClient(httpClient, didReceiveResponse:response)
  }.addDisposableTo(bag)
}
```
to interface with imperative code, here are the options that work well.

If `performRequest` is a finite sequence, and one doesn't care about cancelling computation on deinit, then you don't need `disposeBag`:

```swift
func request() {
  let httpClient = HttpClient()
  _ = httpClient.performRequest().single().subscribeNext { [weak self] response in
    self?.httpClient(httpClient, didReceiveResponse:response)
  }
}
```

... or if you want to cancel request when deallocated, but don't care if there are multiple parallel in flight requests:

```swift
func request() {
  let httpClient = HttpClient()
  _ = httpClient.performRequest().single().takeUntil(self.rx.deallocated).subscribeNext { [weak self] response in
    self?.httpClient(httpClient, didReceiveResponse:response)
  }
}
```

... or a more complete solution if one wants to cancel computation when object deallocates and make sure that there is only one request in flight at any time:

```swift
class Object {
    let disposeBag: DisposeBag
    let makeRequest = PublishSubject()
    init() {
        makeRequest.flatMapLatest {
            let httpClient = HttpClient()
            return httpClient.performRequest().single()
        }
        .subscribe(onNext: { [weak self] response
            // do something with response
        })
        .addDisposable(to: bag)
    }

    func request() {
        makeRequest.onNext()
    }
}
```

**Comment by sirlantis** [comment]
Hey @kzaher - thanks for the update. Coincidentally I upgraded to RxSwift 3.0 beta 1 yesterday (great job!) and ran into the same issue yesterday. Updated to `master` and now it looks great!

Happily removing the ugly `// DO NOT ADD TO DISPOSE BAG - THIS WILL LEAK!` note in my code now. 😉

**Comment by Alexey-Matjuk** [comment]
Hey @kzaher. You wrote:
> If performRequest is a finite sequence, and one doesn't care about cancelling computation on deinit, then you don't need disposeBag:
> 
> func request() {
>   let httpClient = HttpClient()
>   _ = httpClient.performRequest().single().subscribeNext { [weak self] response in
>     self?.httpClient(httpClient, didReceiveResponse:response)
>   }
> }
> 
But then we get a warning then: **Result of call to 'subscribe(onNext:onError:onCompleted:onDisposed:)' is unused**
Having warnings in the project is not acceptable for me :(
Any suggestions?
